tensorflow/core/kernels/eigen_backward_spatial_convolutions.h

```diff
@@ -438,154 +400,100 @@ SpatialConvolutionBackwardKernel(
-  DSizes<TensorIndex, 2> input_dims;
+  // Reshaped out
+  DSizes<TensorIndex, 2> output_dims;
   if (isColMajor) {
-    input_dims[0] = kernelChannels;
-    input_dims[1] = batch * inputRows * inputCols;
+    output_dims[0] = kernelFilters;
+    output_dims[1] = outputRows * outputCols;
+    for (int i = 3; i < NumDims; ++i) {
+      output_dims[1] *= out.dimension(i);
+    }
   } else {
-    input_dims[1] = kernelChannels;
-    input_dims[0] = inputCols * inputRows * batch;
+    output_dims[1] = kernelFilters;
+    output_dims[0] = outputCols * outputRows;
+    for (int i = 0; i < NumDims - 3; ++i) {
+      output_dims[0] *= out.dimension(i);
+    }
   }
 
-  // Molds the output of the patch extraction result into a 2D tensor:
-  // - the first dimension (dims[0]): the patch values to be multiplied with the
-  // kernels
-  // - the second dimension (dims[1]): everything else
+  // Reshaped extract_image_patches(in)
   DSizes<TensorIndex, 2> pre_contract_dims;
   if (isColMajor) {
-    pre_contract_dims[0] = batch * inputRows * inputCols;
-    pre_contract_dims[1] = kernelRows * kernelCols * kernelFilters;
+    pre_contract_dims[0] = kernelChannels * kernelRows * kernelCols;
+    pre_contract_dims[1] = outputRows * outputCols;
+    for (int i = 3; i < NumDims; ++i) {
+      pre_contract_dims[1] *= in.dimension(i);
+    }
+    eigen_assert(output_dims[1] == pre_contract_dims[1]);
   } else {
-    pre_contract_dims[1] = inputCols * inputRows * batch;
-    pre_contract_dims[0] = kernelFilters * kernelCols * kernelRows;
+    pre_contract_dims[1] = kernelCols * kernelRows * kernelChannels;
+    pre_contract_dims[0] = outputRows * outputCols;
+    for (int i = 0; i < NumDims - 3; ++i) {
+      pre_contract_dims[0] *= in.dimension(i);
+    }
+    eigen_assert(output_dims[0] == pre_contract_dims[0]);
   }
-
   // We will contract along the collapsed dimension that contains the
-  // batch, inputRows and inputCols.
+  // outputCols, outputRows and OTHERS.
   array<IndexPair<TensorIndex>, 1> contract_dims;
-  contract_dims[0] = IndexPair<TensorIndex>(1, 0);
-
-  // Dimensions after contraction.
-  DSizes<TensorIndex, NumDims> post_contract_dims;
+  // The order of contraction is reversed, as it is much faster !
   if (isColMajor) {
-    post_contract_dims[0] = kernelChannels;
-    post_contract_dims[1] = kernelRows;
-    post_contract_dims[2] = kernelCols;
-    post_contract_dims[3] = kernelFilters;
-  } else {
-    post_contract_dims[0] = kernelFilters;
-    post_contract_dims[1] = kernelCols;
-    post_contract_dims[2] = kernelRows;
-    post_contract_dims[3] = kernelChannels;
-  }
-
-  // Reorder output of contraction to a valid filter shape.
-  array<TensorIndex, 4> kernel_shuffle;
-  if (isColMajor) {
-    // From: [in_depth, kernel_rows, kernel_cols, out_depth]
-    // To:   [out_depth, in_depth, kernel_rows, kernel_cols]
-    kernel_shuffle = {3, 0, 1, 2};
+    // row-major: input.patches.contract(output_backward)
```

Comment:
nit: our style guide requires capitalization and punctuation, here and throughout.

Reply:
In the comments ? As this part is almost completely from old tf 1.13, I might overlooked some comments. Will fix.